drivers/usb/bc12/bc12_pi3usb9201.c

```diff
@@ -148,22 +148,49 @@ static void pi3usb9201_notify_callback(const struct device *dev,
 	}
 }
 
+static bool pi3usb9201_partner_has_changed(const struct device *dev,
+					   struct bc12_partner_state *state)
+{
+	struct pi3usb9201_data *pi3usb9201_data = dev->data;
+
+	/* Always notify when clearing out partner state */
+	if (!state) {
+		return true;
+	}
+
+	if (state->bc12_role != pi3usb9201_data->partner_state.bc12_role) {
+		return true;
+	}
+
+	if (state->bc12_role == BC12_PORTABLE_DEVICE &&
+	    pi3usb9201_data->partner_state.type != state->type) {
+		return true;
+	}
+
+	if (state->bc12_role == BC12_CHARGING_PORT &&
+	    pi3usb9201_data->partner_state.pd_partner_connected != state->pd_partner_connected) {
+		return true;
+	}
+
+	return false;
+}
+
 static void pi3usb9201_update_charging_partner(const struct device *dev,
 					       struct bc12_partner_state *const state)
 {
 	struct pi3usb9201_data *pi3usb9201_data = dev->data;
 
-	if (state && state->type == pi3usb9201_data->partner_type) {
-		/* No change to the charging partner */
+	if (!pi3usb9201_partner_has_changed(dev, state)) {
+		/* No change to the partner */
 		return;
 	}
 
-	if (state && state->type != BC12_TYPE_NONE) {
-		/* Now update the current charger type */
-		pi3usb9201_data->partner_type = state->type;
+	if (state) {
+		/* Now update callback with the new partner type. */
+		pi3usb9201_data->partner_state = *state;
 		pi3usb9201_notify_callback(dev, state);
 	} else {
-		pi3usb9201_data->partner_type = BC12_TYPE_NONE;
+		pi3usb9201_data->partner_state.bc12_role = BC12_DISCONNECTED;
```

Comment:
I found the use of the state parameter a bit confusing, NULL is passed only in pi3usb9201_disconnect() to cleanup, right?

Reply:
Added a description of the function and parameters.